### ISSUE TITLE
Investigate to use `thiserror` in our crates

### ISSUE THREAD

**Issue by utam0k** [containers/youki#1377]
# Introduction

Now we are using `anyhow` everywhere. But It would be better to provide some codes as crates(e.g. libcontainer, libcgroup...). Their crates in their current state are not very user-friendly of error handling.

# Goal

Investigate and implement to use of `thiserror` instead of `anyhow` in crates

# Tracking

- [ ] [liboci-cli](https://github.com/containers/youki/tree/main/crates/liboci-cli)
- [ ] [libcgroups](https://github.com/containers/youki/tree/main/crates/libcgroups)
- [ ] [libcontainer](https://github.com/containers/youki/tree/main/crates/libcontainer)

**Comment by utam0k** [comment]
@chermehdi I'm happy if you have interested in this issue. But this issue is not easy and take a lof of time to achieve better than other good first issues.

**Comment by chermehdi** [comment]
Sounds good, let us do them in that order, the first one should give us a pretty good idea about how will the others look like, and how much effort is required as well.

**Comment by utam0k** [comment]
@chermehdi That's right, libcontainer is going to be quite large task, so if there is a way to make it divisible that would be nice too. May I assign you? There is no deadline, etc. Of course, I can help you whenever.

**Comment by chermehdi** [comment]
@utam0k yeah sounds good, just assing me.

**Comment by chermehdi** [comment]
@utam0k `liboci-cli` only contains arg definitions for the cli implementation. I don't see any behavior that requires error handling, I am not sure why it was included as part of this issue. Can you clarify?

**Comment by utam0k** [comment]
> @utam0k `liboci-cli` only contains arg definitions for the cli implementation. I don't see any behavior that requires error handling, I am not sure why it was included as part of this issue. Can you clarify?

Ops, sorry. You don't need to take it. Let's look into libcgroups.

**Comment by chermehdi** [comment]
@utam0k Where do we want to use thiserror? I suggest to minimize the changes we only introduce thiserror usage right at the edge of the library and keep error handling internally using anyhow/context. this will minimize the number of changes we have to make.

Another thing that I want us to think about is what is our API boundary, and how we want this to be used as a library.

* Do we want to provide libcgroups as it's own crate? or will we provide Youki as a library as a whole with the API like `createFromSpec`, `start`, `stop`?

**Comment by utam0k** [comment]
@chermehdi Sorry for the late reply.

:100: 
> I suggest to minimize the changes we only introduce thiserror usage right at the edge of the library and keep error handling internally using anyhow/context. t

Ideally, yes. Once this is possible, it can be spun out and offered like [containers/oci-spec-rs](https://github.com/containers/oci-spec-rs).
> Do we want to provide libcgroups as it's own crate?

**Comment by squili** [comment]
Has there been any progress on this? If not, I'd be happy to help!

**Comment by utam0k** [comment]
👋 Hi, @squili. Unfortunately, we don't have any progress. I would like to introduce thiserror with libcgroup and libcontainer. Are you interested?

**Comment by squili** [comment]
Sure, I'd love to help with that!

**Comment by utam0k** [comment]
Which crate would you like to choose?

**Comment by squili** [comment]
Looks like libcgroup needs to be first, since it's a dependency of libcontainer

**Comment by utam0k** [comment]
I have given you an issue. If possible, could you please generate a plan and create a list of tasks to be done? This can be made later. Thank you.

**Comment by utam0k** [comment]
@chermehdi If you have made any progress, please let me know. I know you are probably very busy.

**Comment by squili** [comment]
Here's a quick plan idea for the conversion, based on the structure of the standard library:

- [ ] Add `thiserror`
- [ ] Create folder-level error enums
  - e.g. `v2::Error` for libcgroups::v2
  - See `std::io::Error`
  - Add initial `Anyhow(#[from] anyhow::Error)` variant
- [ ] Add `Error` type to traits
  - See `core::str::FromStr`
  - Set module errors on the impls
- [ ] Add enum variants until the `Anyhow` variant can be removed
  - This will initially be messy, which is okay!
- [ ] Remove `anyhow`
- [ ] Refactor out smaller error types
  - See `std::string::FromUtf8Error` for example of where a struct error can be helpful
  - Will be able to see where folder-level variants are only used once or only in one module, which will help figure out which variants to refactor out

After this, there should be a "good-enough" error structure, which can be iterated on.

**Comment by utam0k** [comment]
@squili Do you want to create a feature branch?

**Comment by squili** [comment]
I can make a branch on a fork and PR

**Comment by squili** [comment]
Finished work for libcgroups, ready to review! It still uses anyhow in tests because it's useful for tests, but it's been removed as a dependency for everything else.

**Comment by yihuaf** [comment]
@squili Sorry I did not catch this thread. I made #1876 and hope I did not step on your toes.

**Comment by utam0k** [comment]
@yihuaf May I ask you to create TODOs list and share them with @squili ?

**Comment by yihuaf** [comment]
Here we go.

- [ ] container
- [ ] process (partly done)
- [ ] rootfs
- [ ] seccomp
- [ ] workload
- [ ] apparmor
- [ ] capabilities
- [ ] config
- [ ] hook
- [ ] notify_socket
- [ ] signal
- [ ] tty
- [ ] utils

At the top level, we should have a concrete error type in `lib.rs` to make `libcontainer` a stand alone library.

Finally

- [ ] remove anyhow

@squili Let's just leave a note in this thread to coordinate :)

**Comment by squili** [comment]
@yihuaf ill make sure to leave a note for the things i work on! though im going through a few things right now so it may be a bit

**Comment by yihuaf** [comment]
I plan on to work on these today:

capabilities
config
hook
notify_socket
signal

**Comment by YJDoc2** [comment]
@yihuaf and @squili , thanks for your efforts! May I ask you to add the related PR links (existing and future) in the issue description itself? I think it will help in keeping track of which all PRs were part of this change.
